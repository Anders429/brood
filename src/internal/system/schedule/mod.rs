pub(crate) mod raw_task;
pub(crate) mod task;
