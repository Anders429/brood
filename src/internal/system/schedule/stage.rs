use crate::{
    internal::system::schedule::sendable::SendableWorld,
    registry::Registry,
    system::{
        schedule::stage::{Null, Stage},
        ParSystem, System,
    },
};

pub trait StagesSeal<'a>: Send {
    fn run<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;

    fn defer<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;

    fn run_current<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;

    fn run_continuing<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;

    fn flush<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;
}

impl<'a> StagesSeal<'a> for Null {
    fn run<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }

    fn defer<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }

    fn run_current<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }

    fn run_continuing<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }

    fn flush<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }
}

impl<'a, S, P, L> StagesSeal<'a> for (Stage<S, P>, L)
where
    S: System<'a> + Send,
    P: ParSystem<'a> + Send,
    L: StagesSeal<'a>,
{
    fn run<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        self.defer(world);
        self.run_current(world);
    }

    fn defer<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        match self.0 {
            Stage::Start(_) | Stage::Flush => {
                self.1.run(world);
            }
            Stage::Continue(_) => {
                self.1.defer(world);
            }
        }
    }

    fn run_current<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        match &mut self.0 {
            Stage::Start(task) => {
                task.run(world);
            }
            Stage::Continue(task) => {
                rayon::join(
                    || {
                        self.1.run_continuing(world);
                    },
                    || {
                        task.run(world);
                    },
                );
            }
            Stage::Flush => {
                self.1.flush(world);
            }
        }
    }

    fn run_continuing<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        match &mut self.0 {
            Stage::Start(task) => {
                task.run(world);
            }
            Stage::Continue(task) => {
                rayon::join(
                    || {
                        self.1.run_continuing(world);
                    },
                    || {
                        task.run(world);
                    },
                );
            }
            Stage::Flush => {}
        }
    }

    fn flush<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        match &mut self.0 {
            Stage::Start(task) => {
                task.flush(world);
            }
            Stage::Continue(task) => {
                self.1.flush(world);
                task.flush(world);
            }
            Stage::Flush => {}
        }
    }
}
