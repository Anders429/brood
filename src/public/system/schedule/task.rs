use crate::{
    internal::system::schedule::sendable::SendableWorld,
    registry::Registry,
    system::{ParSystem, System},
    world::World,
};

pub enum Task<S, P> {
    Seq(S),
    Par(P),
}

impl<'a, S, P> Task<S, P>
where
    S: System<'a>,
    P: ParSystem<'a>,
{
    pub(crate) fn run<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry,
    {
        match self {
            Task::Seq(system) => {
                // Query world using system.
                let result = unsafe { world.0.query_unchecked::<S::Views, S::Filter>() };
                // Run system using the query result.
                system.run(result);
            }
            Task::Par(system) => {
                // Query world using system.
                let result = unsafe { world.0.par_query_unchecked::<P::Views, P::Filter>() };
                // Run system using the query result.
                system.run(result);
            }
        }
    }

    pub(crate) fn flush<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry,
    {
        let mut_world = unsafe { &mut *(world.0 as *const World<R> as *mut World<R>) };

        match self {
            Task::Seq(system) => system.world_post_processing(mut_world),
            Task::Par(system) => system.world_post_processing(mut_world),
        }
    }
}
