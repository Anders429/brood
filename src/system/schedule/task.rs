use crate::{
    hlist::Null,
    query::{
        view::{
            self,
            Claims,
        },
        Query,
    },
    registry,
    resource,
    system::{
        schedule::{
            stage::SendPtr,
            Stage,
        },
        ParSystem,
        System,
    },
    world::World,
};

pub enum Task<System, ParSystem> {
    System(System),
    ParSystem(ParSystem),
}

impl<System, ParSystem> Task<System, ParSystem> {
    pub(crate) fn run<
        'a,
        Registry,
        Resources,
        QueryIndices,
        ResourceViewIndices,
        DisjointIndices,
        EntryIndices,
        ParQueryIndices,
        ParResourceViewIndices,
        ParDisjointIndices,
        ParEntryIndices,
    >(
        &mut self,
        world: SendPtr<World<Registry, Resources>>,
    ) where
        Registry: registry::ContainsQuery<'a, System::Filter, System::Views<'a>, QueryIndices>
            + registry::ContainsViews<'a, System::EntryViews<'a>, EntryIndices>
            + registry::ContainsParQuery<'a, ParSystem::Filter, ParSystem::Views<'a>, ParQueryIndices>
            + registry::ContainsViews<'a, ParSystem::EntryViews<'a>, ParEntryIndices>
            + 'a,
        Resources: resource::ContainsViews<'a, System::ResourceViews<'a>, ResourceViewIndices>
            + resource::ContainsViews<'a, ParSystem::ResourceViews<'a>, ParResourceViewIndices>
            + 'a,
        System: self::System,
        System::Views<'a>: Send,
        System::ResourceViews<'a>: Send,
        System::EntryViews<'a>: view::Disjoint<System::Views<'a>, Registry, DisjointIndices> + Send,
        ParSystem: self::ParSystem,
        ParSystem::Views<'a>: Send,
        ParSystem::ResourceViews<'a>: Send,
        ParSystem::EntryViews<'a>:
            view::Disjoint<ParSystem::Views<'a>, Registry, ParDisjointIndices> + Send,
    {
        match self {
            Self::System(system) => {
                let result = unsafe {
                    (*world.0).query(Query::<
                        System::Views<'a>,
                        System::Filter,
                        System::ResourceViews<'a>,
                        System::EntryViews<'a>,
                    >::new())
                };
                system.run(result);
            }
            Self::ParSystem(par_system) => {
                let result = unsafe {
                    (*world.0).par_query(Query::<
                        ParSystem::Views<'a>,
                        ParSystem::Filter,
                        ParSystem::ResourceViews<'a>,
                        ParSystem::EntryViews<'a>,
                    >::new())
                };
                par_system.run(result);
            }
        }
    }

    fn claims<
        'a,
        Registry,
        Resources,
        ViewsIndices,
        EntryViewsIndices,
        ResourceViewsIndices,
        ParViewsIndices,
        ParEntryViewsIndices,
        ParResourceViewsIndices,
    >(
        &self,
    ) -> (Registry::Claims, Resources::Claims)
    where
        System: self::System,
        ParSystem: self::ParSystem,
        Registry: registry::ContainsViews<'a, System::Views<'a>, ViewsIndices>
            + registry::ContainsViews<'a, System::EntryViews<'a>, EntryViewsIndices>
            + registry::ContainsViews<'a, ParSystem::Views<'a>, ParViewsIndices>
            + registry::ContainsViews<'a, ParSystem::EntryViews<'a>, ParEntryViewsIndices>,
        Resources: resource::ContainsViews<'a, System::ResourceViews<'a>, ResourceViewsIndices>
            + resource::ContainsViews<'a, ParSystem::ResourceViews<'a>, ParResourceViewsIndices>,
    {
        match self {
            Self::System(_) => (
                unsafe {
                    <Registry as registry::contains::views::Sealed<
                        'a,
                        System::Views<'a>,
                        ViewsIndices,
                    >>::claims()
                    .merge_unchecked(&<Registry as registry::contains::views::Sealed<
                        'a,
                        System::EntryViews<'a>,
                        EntryViewsIndices,
                    >>::claims())
                },
                <Resources as resource::contains::views::Sealed<
                    'a,
                    System::ResourceViews<'a>,
                    ResourceViewsIndices,
                >>::claims(),
            ),
            Self::ParSystem(_) => (
                unsafe {
                    <Registry as registry::contains::views::Sealed<
                        'a,
                        ParSystem::Views<'a>,
                        ParViewsIndices,
                    >>::claims()
                    .merge_unchecked(&<Registry as registry::contains::views::Sealed<
                        'a,
                        ParSystem::EntryViews<'a>,
                        ParEntryViewsIndices,
                    >>::claims())
                },
                <Resources as resource::contains::views::Sealed<
                    'a,
                    ParSystem::ResourceViews<'a>,
                    ParResourceViewsIndices,
                >>::claims(),
            ),
        }
    }
}

/// A heterogeneous list of tasks.
///
/// A task is a system to be run on a [`World`]. A set of tasks implementing this trait can be
/// converted into a set of stages, dividing the tasks into parallelizable groups.
pub trait Tasks<'a, Registry, Resources, Indices>:
    Sealed<'a, Registry, Resources, Indices>
{
}

impl<'a, Tasks, Registry, Resources, Indices> self::Tasks<'a, Registry, Resources, Indices>
    for Tasks
where
    Tasks: Sealed<'a, Registry, Resources, Indices>,
{
}

pub trait Sealed<'a, Registry, Resources, Indices> {
    type Stages;

    fn into_stages(self) -> Self::Stages
    where
        Registry: registry::Claims,
        Resources: resource::Claims;
    fn into_stages_internal(self) -> (Registry::Claims, Resources::Claims, Self::Stages)
    where
        Registry: registry::Claims,
        Resources: resource::Claims;
}

impl<Registry, Resources> Sealed<'_, Registry, Resources, Null> for Null {
    type Stages = Null;

    fn into_stages(self) -> Self::Stages
    where
        Registry: registry::Claims,
        Resources: resource::Claims,
    {
        Null
    }

    fn into_stages_internal(self) -> (Registry::Claims, Resources::Claims, Self::Stages)
    where
        Registry: registry::Claims,
        Resources: resource::Claims,
    {
        (Registry::empty_claims(), Resources::empty_claims(), Null)
    }
}

impl<
        'a,
        Tasks,
        System,
        ParSystem,
        Registry,
        Resources,
        ViewsIndices,
        EntryViewsIndices,
        ResourceViewsIndices,
        ParViewsIndices,
        ParEntryViewsIndices,
        ParResourceViewsIndices,
        Indices,
    >
    Sealed<
        'a,
        Registry,
        Resources,
        (
            ViewsIndices,
            EntryViewsIndices,
            ResourceViewsIndices,
            ParViewsIndices,
            ParEntryViewsIndices,
            ParResourceViewsIndices,
            Indices,
        ),
    > for (Task<System, ParSystem>, Tasks)
where
    Tasks: Sealed<'a, Registry, Resources, Indices>,
    System: self::System,
    ParSystem: self::ParSystem,
    Registry: registry::ContainsViews<'a, System::Views<'a>, ViewsIndices>
        + registry::ContainsViews<'a, System::EntryViews<'a>, EntryViewsIndices>
        + registry::ContainsViews<'a, ParSystem::Views<'a>, ParViewsIndices>
        + registry::ContainsViews<'a, ParSystem::EntryViews<'a>, ParEntryViewsIndices>,
    Resources: resource::ContainsViews<'a, System::ResourceViews<'a>, ResourceViewsIndices>
        + resource::ContainsViews<'a, ParSystem::ResourceViews<'a>, ParResourceViewsIndices>,
{
    type Stages = (Stage<System, ParSystem>, Tasks::Stages);

    fn into_stages(self) -> Self::Stages
    where
        Registry: registry::Claims,
        Resources: resource::Claims,
    {
        self.into_stages_internal().2
    }

    fn into_stages_internal(self) -> (Registry::Claims, Resources::Claims, Self::Stages)
    where
        Registry: registry::Claims,
        Resources: resource::Claims,
    {
        let (component_claims, resource_claims, stages) = self.1.into_stages_internal();
        let (new_component_claims, new_resource_claims) = self.0.claims::<Registry, Resources, ViewsIndices, EntryViewsIndices, ResourceViewsIndices, ParViewsIndices, ParEntryViewsIndices, ParResourceViewsIndices>();
        match (
            component_claims.try_merge(&new_component_claims),
            resource_claims.try_merge(&new_resource_claims),
        ) {
            (Some(component_claims), Some(resource_claims)) => (
                component_claims,
                resource_claims,
                (Stage::Continue(self.0), stages),
            ),
            _ => (
                Registry::empty_claims(),
                Resources::empty_claims(),
                (Stage::Start(self.0), stages),
            ),
        }
    }
}
