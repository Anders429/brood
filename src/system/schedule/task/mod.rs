//! Tasks that are used to define a [`Schedule`].
//!
//! [`Schedule`]: crate::system::Schedule

mod sealed;

pub(crate) use sealed::Task;

use crate::hlist::define_null;

define_null!();

/// A task that implements [`System`].
pub struct System<S>(pub S);

/// A task that implements [`ParSystem`].
pub struct ParSystem<P>(pub P);
