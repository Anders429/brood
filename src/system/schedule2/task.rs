use crate::{hlist::define_null, query::Query, registry::{Registry, ContainsQuery, ContainsParQuery}, system, system::{schedule2::sendable::SendableWorld}};

define_null!();

pub trait Task<R, SFI, SVI, SP, SI, SQ> where R: Registry {
    fn run(&mut self, world: SendableWorld<R>);
}

pub struct System<S>(S);

impl<R, S, SFI, SVI, SP, SI, SQ> Task<R, SFI, SVI, SP, SI, SQ> for System<S> where S: system::System + Send, R: for <'a> ContainsQuery<'a, S::Filter, SFI, S::Views<'a>, SVI, SP, SI, SQ> {
    fn run(&mut self, world: SendableWorld<R>) {
        // Query world using system.
        let result =
            // SAFETY: The access to the world's components follows Rust's borrowing rules.
            unsafe { (*world.get()).query(Query::<S::Views<'_>, S::Filter>::new()) };
        // Run system using the query result.
        self.0.run(result);
    }
}

pub struct ParSystem<P>(P);

impl<P, R, SFI, SVI, SP, SI, SQ> Task<R, SFI, SVI, SP, SI, SQ> for ParSystem<P> where P: system::ParSystem + Send, R: for <'a> ContainsParQuery<'a, P::Filter, SFI, P::Views<'a>, SVI, SP, SI, SQ> {
    fn run(&mut self, world: SendableWorld<R>) {
        // Query world using system.
        let result =
            // SAFETY: The access to the world's components follows Rust's borrowing rules.
            unsafe { (*world.get()).par_query(Query::<P::Views<'_>, P::Filter>::new()) };
        // Run system using the query result.
        self.0.run(result);
    }
}
