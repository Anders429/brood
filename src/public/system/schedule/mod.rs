pub mod stage;
pub mod task;

mod builder;

pub use builder::Builder;

use crate::{
    internal::system::schedule::{sendable::SendableWorld, raw_task}, registry::Registry,
    world::World,
};
use stage::Stages;

pub struct Schedule<S> {
    stages: S,
}

impl Schedule<stage::Null> {
    pub fn builder() -> Builder<raw_task::Null> {
        Builder::new()
    }
}

impl<'a, S> Schedule<S>
where
    S: Stages<'a>,
{
    pub(crate) fn run<R>(&mut self, world: &'a mut World<R>)
    where
        R: Registry,
    {
        self.stages.run(SendableWorld(world));
    }
}
