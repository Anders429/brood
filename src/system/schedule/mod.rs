//! A list of [`System`]s to be run in stages.
//!
//! [`Schedule`]s are created using a builder pattern. `System`s are provided in the desired order
//! they are to be run, and the stages in which those `System`s are run is automatically derived.
//!
//! The advantage of defining a `Schedule` is that `System`s are allowed to be run in parallel as
//! long as their [`Views`] can be borrowed simultaneously.
//!
//! # Example
//! The below example will execute both `SystemA` and `SystemB` in parallel, since their views can
//! be borrowed simultaneously.
//!
//! ``` rust
//! use brood::{
//!     query::{
//!         filter,
//!         filter::Filter,
//!         result,
//!         views,
//!     },
//!     registry::ContainsQuery,
//!     system::{
//!         Schedule,
//!         System,
//!     },
//! };
//!
//! // Define components.
//! struct Foo(usize);
//! struct Bar(bool);
//! struct Baz(f64);
//!
//! struct SystemA;
//!
//! impl<'a> System<'a> for SystemA {
//!     type Views = views!(&'a mut Foo, &'a Bar);
//!     type Filter = filter::None;
//!
//!     fn run<R, FI, VI, P, I, Q>(
//!         &mut self,
//!         query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
//!     ) where
//!         R: ContainsQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
//!     {
//!         for result!(foo, bar) in query_results {
//!             // Do something...
//!         }
//!     }
//! }
//!
//! struct SystemB;
//!
//! impl<'a> System<'a> for SystemB {
//!     type Views = views!(&'a mut Baz, &'a Bar);
//!     type Filter = filter::None;
//!
//!     fn run<R, FI, VI, P, I, Q>(
//!         &mut self,
//!         query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
//!     ) where
//!         R: ContainsQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
//!     {
//!         for result!(baz, bar) in query_results {
//!             // Do something...
//!         }
//!     }
//! }
//!
//! let schedule = Schedule::builder().system(SystemA).system(SystemB).build();
//! ```
//!
//! [`Schedule`]: crate::system::schedule::Schedule
//! [`System`]: crate::system::System
//! [`Views`]: crate::query::view::Views

pub mod raw_task;
pub mod stage;

pub(crate) mod task;

mod sendable;

mod builder;

pub use builder::Builder;
pub use stage::stages;

use crate::{
    registry::Registry,
    world::World,
};
use sendable::SendableWorld;
use stage::Stages;

/// A list of [`System`]s to be run in stages.
///
/// The `System`s that make up a `Schedule` are organized into [`Stages`] on creation. `System`s
/// that can be run in parallel are done so. See the documentation for the [`schedule::Builder`]
/// for more information about creating a `Schedule`.
///
/// # Example
/// The below example will execute both `SystemA` and `SystemB` in parallel, since their views can
/// be borrowed simultaneously.
///
/// ``` rust
/// use brood::{
///     query::{
///         filter,
///         filter::Filter,
///         result,
///         views,
///     },
///     registry::ContainsQuery,
///     system::{
///         Schedule,
///         System,
///     },
/// };
///
/// // Define components.
/// struct Foo(usize);
/// struct Bar(bool);
/// struct Baz(f64);
///
/// struct SystemA;
///
/// impl<'a> System<'a> for SystemA {
///     type Views = views!(&'a mut Foo, &'a Bar);
///     type Filter = filter::None;
///
///     fn run<R, FI, VI, P, I, Q>(
///         &mut self,
///         query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
///     ) where
///         R: ContainsQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
///     {
///         for result!(foo, bar) in query_results {
///             // Do something...
///         }
///     }
/// }
///
/// struct SystemB;
///
/// impl<'a> System<'a> for SystemB {
///     type Views = views!(&'a mut Baz, &'a Bar);
///     type Filter = filter::None;
///
///     fn run<R, FI, VI, P, I, Q>(
///         &mut self,
///         query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
///     ) where
///         R: ContainsQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
///     {
///         for result!(baz, bar) in query_results {
///             // Do something...
///         }
///     }
/// }
///
/// let schedule = Schedule::builder().system(SystemA).system(SystemB).build();
/// ```
///
/// [`schedule::Builder`]: crate::system::schedule::Builder
/// [`Stages`]: crate::system::schedule::stage::Stages
/// [`System`]: crate::system::System
#[cfg_attr(doc_cfg, doc(cfg(feature = "rayon")))]
pub struct Schedule<S> {
    stages: S,
}

impl Schedule<stage::Null> {
    /// Creates a [`schedule::Builder`] to construct a new `Schedule`.
    ///
    /// # Example
    /// ``` rust
    /// use brood::system::Schedule;
    ///
    /// let builder = Schedule::builder();
    /// // Add systems to the builder.
    /// let schedule = builder.build();
    /// ```
    ///
    /// [`schedule::Builder`]: crate::system::schedule::Builder
    #[must_use]
    pub fn builder() -> Builder<raw_task::Null> {
        Builder::new()
    }
}

impl<'a, S> Schedule<S> {
    pub(crate) fn run<R, SFI, SVI, PFI, PVI, SP, SI, SQ, PP, PI, PQ>(
        &mut self,
        world: &'a mut World<R>,
    ) where
        R: Registry,
        S: Stages<'a, R, SFI, SVI, PFI, PVI, SP, SI, SQ, PP, PI, PQ>,
    {
        self.stages.run(
            // SAFETY: The pointer provided here is unique, being created from a mutable reference.
            unsafe { SendableWorld::new(world) },
        );
    }
}
