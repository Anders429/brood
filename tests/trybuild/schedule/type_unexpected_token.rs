use brood::{query::{views, filter, result}, registry::ContainsQuery, system::{System, Schedule}};
// This import is technically unused, since the macro fails to compile before it would be consumed.
// I'm leaving it here, though, for completeness; user code would use this module, and these tests
// should do their best to simulate user code.
#[allow(unused_imports)]
use brood::system::schedule::task;

// Define systems.
struct A;

impl System for A {
    type Views<'a> = views!();
    type Filter = filter::None;

    fn run<'a, R, FI, VI, P, I, Q>(
        &mut self,
        _query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views<'a>, VI, P, I, Q>,
    ) where
        R: ContainsQuery<'a, Self::Filter, FI, Self::Views<'a>, VI, P, I, Q> {}
}

struct B;

impl System for B {
    type Views<'a> = views!();
    type Filter = filter::None;

    fn run<'a, R, FI, VI, P, I, Q>(
        &mut self,
        _query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views<'a>, VI, P, I, Q>,
    ) where
        R: ContainsQuery<'a, Self::Filter, FI, Self::Views<'a>, VI, P, I, Q> {}
}

type MySchedule = Schedule!(task::System<A>, + task::System<B>,);

fn main() {}
