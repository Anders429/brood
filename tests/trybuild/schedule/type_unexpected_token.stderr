error: no rules expected the token `+`
  --> tests/trybuild/schedule/type_unexpected_token.rs:39:44
   |
39 | type Schedule = Schedule!(task::System<A>, + task::System<B>,);
   |                                            ^ no rules expected this token in macro call
   |
note: while trying to match meta-variable `$tasks:ty`
  --> src/system/schedule/mod.rs
   |
   |             ($task:ty $(,$tasks:ty)* $(,)?) => (
   |                          ^^^^^^^^^
